#
# Generated by c:\me\bin\mjm at Thu Jun 27 15:31:01 2013

#all: UniqueStartTracker.class SoftClipTrack.class StringCounter.class SimpleIntMap.class SAMReadCounter.class IndelInfo.class PairedReadWriter.class SNP.class PolyReport.class SAMRecordArrayIterable.class NIB.class Codon.class dbSNPEntry.class AcePanel.class ReferenceOrVariant.class SAMResource.class LittleEndianReader.class dbSNPQuery.class SampleMatchInfo.class MultiplexedWriter.class AceSequence.class RefGene.class Alignment.class SNPControl.class SEUWriterFASTQ.class GBSeq.class SequenceViewLock.class Contig.class SAMAssembly.class PaddedNucleotideSearch.class SAMConsensusBuilder.class BigEndianReader.class FASTAFileProperties.class ReadReport.class FASTAQualityReader.class Sample.class SEUWriterBAM.class ProgressInfo.class SAMReportPositions.class SAMFinneyCoverage.class TumorNormalReferenceTracker.class SAMIndelFilter.class PhredPolyFile.class SAMStreamingSNPFinder.class SearchWidget2.class SAMTracePeakPositions.class SAMCoverage.class BaseCountInfo.class CoordTag.class FASTAIndexedFAI.class BAMReferenceCompatibility.class IntMap.class MemoryMonitor.class BaseCounter3.class AlleleSkew.class TextWindow.class SAMMismatchFrequency.class Exon.class SNP2.class SEUConfig.class BambinoProperties.class SAMReadQuery.class .class SNPTrackInfo.class PagedIntMap.class SAMPooledIterator.class HeteroSummary.class GeneInfo.class AceViewerConfig.class SAMRecordArrayIterator.class SAMConsensusMapFactory.class SAMResourceTags.class RangeMasker.class JDBCCache.class SAMReadQualityAnalysis.class FASTADirectory.class Range.class IntronCompressor.class SAMChromName.class Reporter.class SAMReport.class GenomeBuildInfo.class SAMAmbigCounter.class MetaExonIntronCoverage.class Ruler.class SAMHQMismatchReport.class SAMExtractUnmapped.class dbSNP.class GermlineSomaticLOH.class SampleNamingConvention.class ReferenceSequenceByte.class Filfre.class SAMTagCounter.class Genotyper.class WorkingDirectory.class AceViewer.class PhredPolyFileEntry.class dbSNPSet.class AlignmentSequence.class AnnotationLoader.class StrandSkewFilter.class SplicedReadReporter.class SplicedReadInfo.class LibraryInfo.class FAIIndexRecord.class SAMMismatchFilter.class FontWidthTracker.class SAMResourceIterator.class SuffixFileFilter.class SAMRegion.class CDSChangePredictor.class KnownGenomeBuild.class ClipboardSetter.class BAMConcat.class AssemblySequence.class UniqueReadName.class SEUInteresting.class dbSNPDB.class SAMCache.class GZIPOutputStreamEx.class SpinnerNumber.class ArrowPainter.class ChromosomeDisambiguator.class FASTASequenceReader.class TwoBitFile.class SAMUtils.class SAMDuplicateReport.class SAMCanonicalReference.class SampleNamingConventionPattern.class IntMapBenchmark.class ColorIntensifier.class HashList.class MarkupReader.class SAMQuery.class NIOUtils.class SAMConsensusMapping.class TumorNormal.class SAMTagFilter.class SNPList.class ColorParser.class Ace.class MapQTracker.class Base.class IntBucketIndex.class TranscriptExonIntronCoverage.class Chromosome.class PadMap.class PairedExtractHack.class SAMReadLength.class TwoBitSequence.class SAMTagTracker.class DuplicateMapReport.class GeneList.class ReferenceSequence.class TwoBitHeader.class Assembly.class CDSChangeInfo.class ProteinTool.class SAMAdapterTrimmer.class ReadMateExtractorPaired.class Counter.class TwoBitFileLite.class ErrorReporter.class FASTAFileTools.class SortedJobQueue.class Launcher.class HashCounter.class JointPolymorphismScore.class HashCounter2.class TwoBitBlock.class SAMCoverage2.class TraceViewerCache.class AlleleSkewInfo.class SAMConsensusGenerator2.class dbSNPQueryCacher.class ReadMateExtractorUnpaired.class SAMHeaderReport.class EndianReader.class Strand.class SAMConsensusGenerator.class SAMDumpIndels.class SEUWriter.class LongCounter.class SEUMateQuery.class GermlineSomaticTestCase.class AssemblyCanvas.class SAMAlternatePeakAmplitudeRatios.class SAMExtractUnmapped2.class Constants.class TumorNormalStatusDialog.class SNPConfig.class SAMExtractSoftClipped.class WorkingFile.class SearchWidget.class dbSNPHeader.class 


#ExonIntronCoverage.class  rpf
#SplicedReadFlankingInfo.class  rpf
#MetaExonIntronCoverage.class  rpf
#SplicedReadReporter.class  rpf 
#SplicedReadInfo.class  rpf 
#IntronCache.class  rpf
#TranscriptExonIntronCoverage.class  rpf
# ReadMateExtractorPaired.class 
# ReadMateExtractorUnpaired.class  rpf 

all: UniqueStartTracker.class SoftClipTrack.class StringCounter.class SimpleIntMap.class SAMReadCounter.class IndelInfo.class PairedReadWriter.class SNP.class PolyReport.class SAMRecordArrayIterable.class NIB.class Codon.class dbSNPEntry.class AcePanel.class ReferenceOrVariant.class SAMResource.class LittleEndianReader.class dbSNPQuery.class SampleMatchInfo.class MultiplexedWriter.class AceSequence.class RefGene.class Alignment.class SNPControl.class SEUWriterFASTQ.class GBSeq.class SequenceViewLock.class Contig.class SAMAssembly.class PaddedNucleotideSearch.class SAMConsensusBuilder.class BigEndianReader.class FASTAFileProperties.class ReadReport.class FASTAQualityReader.class Sample.class SEUWriterBAM.class ProgressInfo.class SAMReportPositions.class SAMFinneyCoverage.class TumorNormalReferenceTracker.class SAMIndelFilter.class PhredPolyFile.class SAMStreamingSNPFinder.class SearchWidget2.class SAMTracePeakPositions.class SAMCoverage.class BaseCountInfo.class CoordTag.class FASTAIndexedFAI.class BAMReferenceCompatibility.class IntMap.class MemoryMonitor.class BaseCounter3.class AlleleSkew.class TextWindow.class SAMMismatchFrequency.class Exon.class SNP2.class SEUConfig.class BambinoProperties.class SAMReadQuery.class SNPTrackInfo.class PagedIntMap.class SAMPooledIterator.class HeteroSummary.class GeneInfo.class AceViewerConfig.class SAMRecordArrayIterator.class SAMConsensusMapFactory.class SAMResourceTags.class RangeMasker.class JDBCCache.class SAMReadQualityAnalysis.class FASTADirectory.class Range.class IntronCompressor.class SAMChromName.class Reporter.class SAMReport.class GenomeBuildInfo.class SAMAmbigCounter.class Ruler.class SAMHQMismatchReport.class SAMExtractUnmapped.class dbSNP.class GermlineSomaticLOH.class SampleNamingConvention.class ReferenceSequenceByte.class Filfre.class SAMTagCounter.class Genotyper.class WorkingDirectory.class AceViewer.class PhredPolyFileEntry.class dbSNPSet.class AlignmentSequence.class AnnotationLoader.class StrandSkewFilter.class LibraryInfo.class FAIIndexRecord.class SAMMismatchFilter.class FontWidthTracker.class SAMResourceIterator.class SuffixFileFilter.class SAMRegion.class CDSChangePredictor.class KnownGenomeBuild.class ClipboardSetter.class BAMConcat.class AssemblySequence.class UniqueReadName.class SEUInteresting.class dbSNPDB.class SAMCache.class GZIPOutputStreamEx.class SpinnerNumber.class ArrowPainter.class ChromosomeDisambiguator.class FASTASequenceReader.class TwoBitFile.class SAMUtils.class SAMDuplicateReport.class SAMCanonicalReference.class SampleNamingConventionPattern.class IntMapBenchmark.class ColorIntensifier.class HashList.class MarkupReader.class SAMQuery.class NIOUtils.class SAMConsensusMapping.class TumorNormal.class SAMTagFilter.class SNPList.class ColorParser.class Ace.class MapQTracker.class Base.class IntBucketIndex.class Chromosome.class PadMap.class PairedExtractHack.class SAMReadLength.class TwoBitSequence.class SAMTagTracker.class DuplicateMapReport.class GeneList.class ReferenceSequence.class TwoBitHeader.class Assembly.class CDSChangeInfo.class ProteinTool.class SAMAdapterTrimmer.class Counter.class TwoBitFileLite.class ErrorReporter.class FASTAFileTools.class SortedJobQueue.class Launcher.class HashCounter.class JointPolymorphismScore.class HashCounter2.class TwoBitBlock.class SAMCoverage2.class TraceViewerCache.class AlleleSkewInfo.class SAMConsensusGenerator2.class dbSNPQueryCacher.class SAMHeaderReport.class EndianReader.class Strand.class SAMConsensusGenerator.class SAMDumpIndels.class SEUWriter.class LongCounter.class SEUMateQuery.class GermlineSomaticTestCase.class AssemblyCanvas.class SAMAlternatePeakAmplitudeRatios.class SAMExtractUnmapped2.class Constants.class TumorNormalStatusDialog.class SNPConfig.class SAMExtractSoftClipped.class WorkingFile.class SearchWidget.class dbSNPHeader.class 


Ace.class: Ace.java
	javac -deprecation  $?

AcePanel.class: AcePanel.java
	javac -deprecation  $?

AceSequence.class: AceSequence.java
	javac -deprecation  $?

AceViewer.class: AceViewer.java
	javac -deprecation  $?

AceViewerConfig.class: AceViewerConfig.java
	javac -deprecation  $?

Alignment.class: Alignment.java
	javac -deprecation  $?

AlignmentSequence.class: AlignmentSequence.java
	javac -deprecation  $?

AlleleSkew.class: AlleleSkew.java
	javac -deprecation  $?

AlleleSkewInfo.class: AlleleSkewInfo.java
	javac -deprecation  $?

AnnotationLoader.class: AnnotationLoader.java
	javac -deprecation  $?

ArrowPainter.class: ArrowPainter.java
	javac -deprecation  $?

Assembly.class: Assembly.java
	javac -deprecation  $?

AssemblyCanvas.class: AssemblyCanvas.java
	javac -deprecation  $?

AssemblySequence.class: AssemblySequence.java
	javac -deprecation  $?

BambinoProperties.class: BambinoProperties.java
	javac -deprecation  $?

BAMConcat.class: BAMConcat.java
	javac -deprecation  $?

BAMReferenceCompatibility.class: BAMReferenceCompatibility.java
	javac -deprecation  $?

Base.class: Base.java
	javac -deprecation  $?

BaseCounter3.class: BaseCounter3.java
	javac -deprecation  $?

BaseCountInfo.class: BaseCountInfo.java
	javac -deprecation  $?

BigEndianReader.class: BigEndianReader.java
	javac -deprecation  $?

CDSChangeInfo.class: CDSChangeInfo.java
	javac -deprecation  $?

CDSChangePredictor.class: CDSChangePredictor.java
	javac -deprecation  $?

Chromosome.class: Chromosome.java
	javac -deprecation  $?

ChromosomeDisambiguator.class: ChromosomeDisambiguator.java
	javac -deprecation  $?

ClipboardSetter.class: ClipboardSetter.java
	javac -deprecation  $?

Codon.class: Codon.java
	javac -deprecation  $?

ColorIntensifier.class: ColorIntensifier.java
	javac -deprecation  $?

ColorParser.class: ColorParser.java
	javac -deprecation  $?

Constants.class: Constants.java
	javac -deprecation  $?

Contig.class: Contig.java
	javac -deprecation  $?

CoordTag.class: CoordTag.java
	javac -deprecation  $?

Counter.class: Counter.java
	javac -deprecation  $?

dbSNP.class: dbSNP.java
	javac -deprecation  $?

dbSNPDB.class: dbSNPDB.java
	javac -deprecation  $?

dbSNPEntry.class: dbSNPEntry.java
	javac -deprecation  $?

dbSNPHeader.class: dbSNPHeader.java
	javac -deprecation  $?

dbSNPQuery.class: dbSNPQuery.java
	javac -deprecation  $?

dbSNPQueryCacher.class: dbSNPQueryCacher.java
	javac -deprecation  $?

dbSNPSet.class: dbSNPSet.java
	javac -deprecation  $?

DuplicateMapReport.class: DuplicateMapReport.java
	javac -deprecation  $?

EndianReader.class: EndianReader.java
	javac -deprecation  $?

ErrorReporter.class: ErrorReporter.java
	javac -deprecation  $?

Exon.class: Exon.java
	javac -deprecation  $?

#rpf 2013
# ExonIntronCoverage.class: ExonIntronCoverage.java
#	javac -deprecation  $?
#rpf 2013

FAIIndexRecord.class: FAIIndexRecord.java
	javac -deprecation  $?

FASTADirectory.class: FASTADirectory.java
	javac -deprecation  $?

FASTAFileProperties.class: FASTAFileProperties.java
	javac -deprecation  $?

FASTAFileTools.class: FASTAFileTools.java
	javac -deprecation  $?

FASTAIndexedFAI.class: FASTAIndexedFAI.java
	javac -deprecation  $?

FASTAQualityReader.class: FASTAQualityReader.java
	javac -deprecation  $?

FASTASequenceReader.class: FASTASequenceReader.java
	javac -deprecation  $?

Filfre.class: Filfre.java
	javac -deprecation  $?

FontWidthTracker.class: FontWidthTracker.java
	javac -deprecation  $?

GBSeq.class: GBSeq.java
	javac -deprecation  $?

GeneInfo.class: GeneInfo.java
	javac -deprecation  $?

GeneList.class: GeneList.java
	javac -deprecation  $?

GenomeBuildInfo.class: GenomeBuildInfo.java
	javac -deprecation  $?

Genotyper.class: Genotyper.java
	javac -deprecation  $?

GermlineSomaticLOH.class: GermlineSomaticLOH.java
	javac -deprecation  $?

GermlineSomaticTestCase.class: GermlineSomaticTestCase.java
	javac -deprecation  $?

GZIPOutputStreamEx.class: GZIPOutputStreamEx.java
	javac -deprecation  $?

HashCounter.class: HashCounter.java
	javac -deprecation  $?

HashCounter2.class: HashCounter2.java
	javac -deprecation  $?

HashList.class: HashList.java
	javac -deprecation  $?

HeteroSummary.class: HeteroSummary.java
	javac -deprecation  $?

IndelInfo.class: IndelInfo.java
	javac -deprecation  $?

IntBucketIndex.class: IntBucketIndex.java
	javac -deprecation  $?

IntMap.class: IntMap.java
	javac -deprecation  $?

IntMapBenchmark.class: IntMapBenchmark.java
	javac -deprecation  $?

#rpf IntronCache.class: IntronCache.java
#rpf 	javac -deprecation  $?

IntronCompressor.class: IntronCompressor.java
	javac -deprecation  $?

JDBCCache.class: JDBCCache.java
	javac -deprecation  $?

JointPolymorphismScore.class: JointPolymorphismScore.java
	javac -deprecation  $?

KnownGenomeBuild.class: KnownGenomeBuild.java
	javac -deprecation  $?

Launcher.class: Launcher.java
	javac -deprecation  $?

LibraryInfo.class: LibraryInfo.java
	javac -deprecation  $?

LittleEndianReader.class: LittleEndianReader.java
	javac -deprecation  $?

LongCounter.class: LongCounter.java
	javac -deprecation  $?

MapQTracker.class: MapQTracker.java
	javac -deprecation  $?

MarkupReader.class: MarkupReader.java
	javac -deprecation  $?

MemoryMonitor.class: MemoryMonitor.java
	javac -deprecation  $?

#rpf MetaExonIntronCoverage.class: MetaExonIntronCoverage.java
#rpf 	javac -deprecation  $?

MultiplexedWriter.class: MultiplexedWriter.java
	javac -deprecation  $?

NIB.class: NIB.java
	javac -deprecation  $?

NIOUtils.class: NIOUtils.java
	javac -deprecation  $?

PaddedNucleotideSearch.class: PaddedNucleotideSearch.java
	javac -deprecation  $?

PadMap.class: PadMap.java
	javac -deprecation  $?

PagedIntMap.class: PagedIntMap.java
	javac -deprecation  $?

PairedExtractHack.class: PairedExtractHack.java
	javac -deprecation  $?

PairedReadWriter.class: PairedReadWriter.java
	javac -deprecation  $?

PhredPolyFile.class: PhredPolyFile.java
	javac -deprecation  $?

PhredPolyFileEntry.class: PhredPolyFileEntry.java
	javac -deprecation  $?

PolyReport.class: PolyReport.java
	javac -deprecation  $?

ProgressInfo.class: ProgressInfo.java
	javac -deprecation  $?

ProteinTool.class: ProteinTool.java
	javac -deprecation  $?

Range.class: Range.java
	javac -deprecation  $?

RangeMasker.class: RangeMasker.java
	javac -deprecation  $?

#rpf ReadMateExtractorPaired.class: ReadMateExtractorPaired.java
#	javac -deprecation  $?

#ReadMateExtractorUnpaired.class: ReadMateExtractorUnpaired.java
#	javac -deprecation  $?

ReadReport.class: ReadReport.java
	javac -deprecation  $?

ReferenceOrVariant.class: ReferenceOrVariant.java
	javac -deprecation  $?

ReferenceSequence.class: ReferenceSequence.java
	javac -deprecation  $?

ReferenceSequenceByte.class: ReferenceSequenceByte.java
	javac -deprecation  $?

RefGene.class: RefGene.java
	javac -deprecation  $?

Reporter.class: Reporter.java
	javac -deprecation  $?

Ruler.class: Ruler.java
	javac -deprecation  $?

SAMAdapterTrimmer.class: SAMAdapterTrimmer.java
	javac -deprecation  $?

SAMAlternatePeakAmplitudeRatios.class: SAMAlternatePeakAmplitudeRatios.java
	javac -deprecation  $?

SAMAmbigCounter.class: SAMAmbigCounter.java
	javac -deprecation  $?

SAMAssembly.class: SAMAssembly.java
	javac -deprecation  $?

SAMCache.class: SAMCache.java
	javac -deprecation  $?

SAMCanonicalReference.class: SAMCanonicalReference.java
	javac -deprecation  $?

SAMChromName.class: SAMChromName.java
	javac -deprecation  $?

SAMConsensusBuilder.class: SAMConsensusBuilder.java
	javac -deprecation  $?

SAMConsensusGenerator.class: SAMConsensusGenerator.java
	javac -deprecation  $?

SAMConsensusGenerator2.class: SAMConsensusGenerator2.java
	javac -deprecation  $?

SAMConsensusMapFactory.class: SAMConsensusMapFactory.java
	javac -deprecation  $?

SAMConsensusMapping.class: SAMConsensusMapping.java
	javac -deprecation  $?

SAMCoverage.class: SAMCoverage.java
	javac -deprecation  $?

SAMCoverage2.class: SAMCoverage2.java
	javac -deprecation  $?

SAMDumpIndels.class: SAMDumpIndels.java
	javac -deprecation  $?

SAMDuplicateReport.class: SAMDuplicateReport.java
	javac -deprecation  $?

SAMExtractSoftClipped.class: SAMExtractSoftClipped.java
	javac -deprecation  $?

SAMExtractUnmapped.class: SAMExtractUnmapped.java
	javac -deprecation  $?

SAMExtractUnmapped2.class: SAMExtractUnmapped2.java
	javac -deprecation  $?

SAMFinneyCoverage.class: SAMFinneyCoverage.java
	javac -deprecation  $?

SAMHeaderReport.class: SAMHeaderReport.java
	javac -deprecation  $?

SAMHQMismatchReport.class: SAMHQMismatchReport.java
	javac -deprecation  $?

SAMIndelFilter.class: SAMIndelFilter.java
	javac -deprecation  $?

SAMMismatchFilter.class: SAMMismatchFilter.java
	javac -deprecation  $?

SAMMismatchFrequency.class: SAMMismatchFrequency.java
	javac -deprecation  $?

Sample.class: Sample.java
	javac -deprecation  $?

SampleMatchInfo.class: SampleMatchInfo.java
	javac -deprecation  $?

SampleNamingConvention.class: SampleNamingConvention.java
	javac -deprecation  $?

SampleNamingConventionPattern.class: SampleNamingConventionPattern.java
	javac -deprecation  $?

SAMPooledIterator.class: SAMPooledIterator.java
	javac -deprecation  $?

SAMQuery.class: SAMQuery.java
	javac -deprecation  $?

SAMReadCounter.class: SAMReadCounter.java
	javac -deprecation  $?

SAMReadLength.class: SAMReadLength.java
	javac -deprecation  $?

SAMReadQualityAnalysis.class: SAMReadQualityAnalysis.java
	javac -deprecation  $?

SAMReadQuery.class: SAMReadQuery.java
	javac -deprecation  $?

SAMRecordArrayIterable.class: SAMRecordArrayIterable.java
	javac -deprecation  $?

SAMRecordArrayIterator.class: SAMRecordArrayIterator.java
	javac -deprecation  $?

SAMRegion.class: SAMRegion.java
	javac -deprecation  $?

SAMReport.class: SAMReport.java
	javac -deprecation  $?

SAMReportPositions.class: SAMReportPositions.java
	javac -deprecation  $?

SAMResource.class: SAMResource.java
	javac -deprecation  $?

SAMResourceIterator.class: SAMResourceIterator.java
	javac -deprecation  $?

SAMResourceTags.class: SAMResourceTags.java
	javac -deprecation  $?

SAMStreamingSNPFinder.class: SAMStreamingSNPFinder.java
	javac -deprecation  $?

SAMTagCounter.class: SAMTagCounter.java
	javac -deprecation  $?

SAMTagFilter.class: SAMTagFilter.java
	javac -deprecation  $?

SAMTagTracker.class: SAMTagTracker.java
	javac -deprecation  $?

SAMTracePeakPositions.class: SAMTracePeakPositions.java
	javac -deprecation  $?

SAMUtils.class: SAMUtils.java
	javac -deprecation  $?

SearchWidget.class: SearchWidget.java
	javac -deprecation  $?

SearchWidget2.class: SearchWidget2.java
	javac -deprecation  $?

SequenceViewLock.class: SequenceViewLock.java
	javac -deprecation  $?

SEUConfig.class: SEUConfig.java
	javac -deprecation  $?

SEUInteresting.class: SEUInteresting.java
	javac -deprecation  $?

SEUMateQuery.class: SEUMateQuery.java
	javac -deprecation  $?

SEUWriter.class: SEUWriter.java
	javac -deprecation  $?

SEUWriterBAM.class: SEUWriterBAM.java
	javac -deprecation  $?

SEUWriterFASTQ.class: SEUWriterFASTQ.java
	javac -deprecation  $?

SimpleIntMap.class: SimpleIntMap.java
	javac -deprecation  $?

SNP.class: SNP.java
	javac -deprecation  $?

SNP2.class: SNP2.java
	javac -deprecation  $?

SNPConfig.class: SNPConfig.java
	javac -deprecation  $?

SNPControl.class: SNPControl.java
	javac -deprecation  $?

SNPList.class: SNPList.java
	javac -deprecation  $?

SNPTrackInfo.class: SNPTrackInfo.java
	javac -deprecation  $?

SoftClipTrack.class: SoftClipTrack.java
	javac -deprecation  $?

SortedJobQueue.class: SortedJobQueue.java
	javac -deprecation  $?

SpinnerNumber.class: SpinnerNumber.java
	javac -deprecation  $?

#rpf SplicedReadFlankingInfo.class: SplicedReadFlankingInfo.java
#rpf 	javac -deprecation  $?

#SplicedReadInfo.class: SplicedReadInfo.java
#	javac -deprecation  $?

#SplicedReadReporter.class: SplicedReadReporter.java
#	javac -deprecation  $?

Strand.class: Strand.java
	javac -deprecation  $?

StrandSkewFilter.class: StrandSkewFilter.java
	javac -deprecation  $?

StringCounter.class: StringCounter.java
	javac -deprecation  $?

SuffixFileFilter.class: SuffixFileFilter.java
	javac -deprecation  $?

TextWindow.class: TextWindow.java
	javac -deprecation  $?

TraceViewerCache.class: TraceViewerCache.java
	javac -deprecation  $?

#rpf TranscriptExonIntronCoverage.class: TranscriptExonIntronCoverage.java
#rpf 	javac -deprecation  $?

TumorNormal.class: TumorNormal.java
	javac -deprecation  $?

TumorNormalReferenceTracker.class: TumorNormalReferenceTracker.java
	javac -deprecation  $?

TumorNormalStatusDialog.class: TumorNormalStatusDialog.java
	javac -deprecation  $?

TwoBitBlock.class: TwoBitBlock.java
	javac -deprecation  $?

TwoBitFile.class: TwoBitFile.java
	javac -deprecation  $?

TwoBitFileLite.class: TwoBitFileLite.java
	javac -deprecation  $?

TwoBitHeader.class: TwoBitHeader.java
	javac -deprecation  $?

TwoBitSequence.class: TwoBitSequence.java
	javac -deprecation  $?

UniqueReadName.class: UniqueReadName.java
	javac -deprecation  $?

UniqueStartTracker.class: UniqueStartTracker.java
	javac -deprecation  $?

WorkingDirectory.class: WorkingDirectory.java
	javac -deprecation  $?

WorkingFile.class: WorkingFile.java
	javac -deprecation  $?

clean: 
	rm -f *.class

test: all
	java Ace2.AceViewer

